File: code/zato-web-admin/src/zato/admin/static/js/definition/kafka.js

// /////////////////////////////////////////////////////////////////////////////

$.fn.zato.data_table.Kafka = new Class({
    toString: function() {
        var s = '<Kafka id:{0} name:{1} is_active:{2}';
        return String.format(s, this.id ? this.id : '(none)',
                                this.name ? this.name : '(none)',
                                this.is_active ? this.is_active : '(none)');
    }
});

// /////////////////////////////////////////////////////////////////////////////

$(document).ready(function() {
    $('#data-table').tablesorter();
    $.fn.zato.data_table.password_required = true;
    $.fn.zato.data_table.class_ = $.fn.zato.data_table.Kafka;
    $.fn.zato.data_table.new_row_func = $.fn.zato.definition.kafka.data_table.new_row;
    $.fn.zato.data_table.parse();
    $.fn.zato.data_table.setup_forms(['name', 'server_list', 'socket_timeout', 'offset_timeout', 'broker_version']);
})

$.fn.zato.definition.kafka.create = function() {
    $.fn.zato.data_table._create_edit('create', 'Create a new Kafka definition', null);
}

$.fn.zato.definition.kafka.edit = function(id) {
    $.fn.zato.data_table._create_edit('edit', 'Update the Kafka definition', id);
}

$.fn.zato.definition.kafka.data_table.new_row = function(item, data, include_tr) {
    var row = '';

    if(include_tr) {
        row += String.format("<tr id='tr_{0}' class='updated'>", item.id);
    }

    var is_active = item.is_active == true;
    var should_use_zookeeper = item.should_use_zookeeper == true;

    row += "<td class='numbering'>&nbsp;</td>";
    row += "<td class='impexp'><input type='checkbox' /></td>";

    row += String.format('<td>{0}</td>', item.name);
    row += String.format('<td>{0}</td>', is_active ? 'Yes' : 'No');
    row += String.format('<td>{0}</td>', item.server_list);

    row += String.format('<td>{0}</td>', should_use_zookeeper ? 'Yes' : 'No');
    row += String.format('<td>{0}</td>', item.source_address ? item.source_address : $.fn.zato.empty_value);
    row += String.format('<td>{0}</td>', item.broker_version);

    row += String.format('<td>{0}</td>', String.format("<a href=\"javascript:$.fn.zato.definition.kafka.edit('{0}')\">Edit</a>", item.id));
    row += String.format('<td>{0}</td>', String.format("<a href='javascript:$.fn.zato.definition.kafka.delete_({0});'>Delete</a>", item.id));
    row += String.format('<td>{0}</td>', String.format("<a href='javascript:$.fn.zato.data_table.ping({0});'>Ping</a>", item.id));

    row += String.format("<td class='ignore item_id_{0}'>{0}</td>", item.id);
    row += String.format("<td class='ignore'>{0}</td>", is_active);

    // 1 -->
    row += String.format("<td class='ignore'>{0}</td>", item.should_use_zookeeper);
    row += String.format("<td class='ignore'>{0}</td>", item.should_exclude_internal_topics);
    row += String.format("<td class='ignore'>{0}</td>", item.socket_timeout);
    row += String.format("<td class='ignore'>{0}</td>", item.offset_timeout);

    // 2 -->
    row += String.format("<td class='ignore'>{0}</td>", item.is_tls_enabled);
    row += String.format("<td class='ignore'>{0}</td>", item.tls_private_key_file);
    row += String.format("<td class='ignore'>{0}</td>", item.tls_cert_file);
    row += String.format("<td class='ignore'>{0}</td>", item.tls_ca_certs_file);

    // 3 -->
    row += String.format("<td class='ignore'>{0}</td>", item.tls_pem_passphrase);
    row += String.format("<td class='ignore'>{0}</td>", item.source_address);

    if(include_tr) {
        row += '</tr>';
    }

    return row;
}

$.fn.zato.definition.kafka.delete_ = function(id) {
    $.fn.zato.data_table.delete_(id, 'td.item_id_',
        'Kafka definition `{0}` deleted',
        'Are you sure you want to delete Kafka definition `{0}`?',
        true);
}

File: code/zato-web-admin/src/zato/admin/web/views/definition/kafka.py
# -*- coding: utf-8 -*-

"""
Copyright (C) 2019, Zato Source s.r.o. https://zato.io

Licensed under AGPLv3, see LICENSE.txt for terms and conditions.
"""

from __future__ import absolute_import, division, print_function, unicode_literals

# stdlib
import logging

# Zato
from zato.common.api import GENERIC
from zato.admin.web.forms import ChangePasswordForm
from zato.admin.web.forms.definition.kafka import CreateForm, EditForm
from zato.admin.web.views import change_password as _change_password, CreateEdit, Delete as _Delete, Index as _Index, \
     method_allowed, ping_connection
from zato.common.odb.model import GenericConn

# ################################################################################################################################

logger = logging.getLogger(__name__)

# ################################################################################################################################

class Index(_Index):
    method_allowed = 'GET'
    url_name = 'definition-kafka'
    template = 'zato/definition/kafka.html'
    service_name = 'zato.generic.connection.get-list'
    output_class = GenericConn
    paginate = True

    class SimpleIO(_Index.SimpleIO):
        input_required = ('cluster_id', 'type_')
        output_required = ('id', 'name', 'server_list', 'socket_timeout', 'offset_timeout', 'broker_version', 'server_list')
        output_optional = ('is_active',  'should_use_zookeeper', 'should_exclude_internal_topics', 'source_address',
            'is_tls_enabled', 'tls_private_key_file', 'tls_cert_file', 'tls_ca_certs_file', 'tls_pem_passphrase')
        output_repeated = True

    def handle(self):
        return {
            'create_form': CreateForm(req=self.req),
            'edit_form': EditForm(prefix='edit', req=self.req),
            'change_password_form': ChangePasswordForm()
        }

# ################################################################################################################################

class _CreateEdit(CreateEdit):
    method_allowed = 'POST'

    class SimpleIO(CreateEdit.SimpleIO):
        input_required = ('name', 'server_list', 'socket_timeout', 'offset_timeout', 'broker_version', 'server_list')
        input_optional = ('is_active',  'should_use_zookeeper', 'should_exclude_internal_topics', 'source_address',
            'is_tls_enabled', 'tls_private_key_file', 'tls_cert_file', 'tls_ca_certs_file', 'tls_pem_passphrase')

        output_required = 'id', 'name'

    def populate_initial_input_dict(self, initial_input_dict):
        initial_input_dict['type_'] = GENERIC.CONNECTION.TYPE.DEF_KAFKA
        initial_input_dict['is_internal'] = False

        # This definition can be used to create connections in both directions
        initial_input_dict['is_channel'] = True
        initial_input_dict['is_outconn'] = True

        initial_input_dict['sec_use_rbac'] = False
        initial_input_dict['pool_size'] = 1 # Not used but required on input

    def on_after_set_input(self):
        pass

    def success_message(self, item):
        return 'Successfully {} Kafka definition `{}`'.format(self.verb, item.name)

# ################################################################################################################################

class Create(_CreateEdit):
    url_name = 'definition-kafka-create'
    service_name = 'zato.generic.connection.create'

# ################################################################################################################################

class Edit(_CreateEdit):
    url_name = 'definition-kafka-edit'
    form_prefix = 'edit-'
    service_name = 'zato.generic.connection.edit'

# ################################################################################################################################

class Delete(_Delete):
    url_name = 'definition-kafka-delete'
    error_message = 'Could not delete outgoing Kafka definition'
    service_name = 'zato.generic.connection.delete'

# ################################################################################################################################

@method_allowed('POST')
def change_password(req):
    return _change_password(req, 'zato.generic.connection.change-password')

# ################################################################################################################################

@method_allowed('POST')
def ping(req, id, cluster_id):
    return ping_connection(req, 'zato.generic.connection.ping', id, 'Kafka definition')

# ################################################################################################################################
